core/src/main/scala/org/apache/spark/sql/CachedDataFrame.scala

```diff
@@ -128,136 +127,142 @@ class CachedDataFrame(session: SparkSession, queryExecution: QueryExecution,
     }
   }
 
-  private[sql] def reset(): Unit = {
-    if (lastExecRDD ne null) clearPartitions(Seq(lastExecRDD))
-    lastExecRDD = cachedRDD
-  }
-  private lazy val rdd_partitions_ = {
-    val _f = classOf[RDD[_]].getDeclaredField("org$apache$spark$rdd$RDD$$partitions_")
-    _f.setAccessible(true)
-    UnsafeHolder.getUnsafe.objectFieldOffset(_f)
+  private[sql] def duplicate(): CachedDataFrame = {
+    val cdf = new CachedDataFrame(snappySession, queryExecution, queryExecutionString,
+      queryPlanInfo, currentQueryExecutionString, currentQueryPlanInfo, cachedRDD,
+      shuffleDependencies, encoder, rddId, noSideEffects, queryHints, -1L, -1L, -1L)
+    cdf.log_ = log_
+    cdf.levelFlags = levelFlags
+    cdf._boundEnc = _boundEnc
+    cdf._rowConverter = _rowConverter
+    cdf.paramLiterals = paramLiterals
+    cdf
   }
 
-  @tailrec
-  private def clearPartitions(rdd: Seq[RDD[_]]): Unit = {
-    val children = rdd.flatMap(r => if (r != null) {
-      r.dependencies.map {
-        case d: NarrowDependency[_] => d.rdd
-        case s: ShuffleDependency[_, _, _] => s.rdd
+  private def reset(): Unit = clearPartitions(cachedRDD :: Nil)
+
+  private def applyCurrentLiterals(): Unit = {
+    if (paramLiterals.length > 0 && (paramLiterals ne currentLiterals)) {
+      for (pos <- paramLiterals.indices) {
+        paramLiterals(pos).updateValue(currentLiterals(pos).value)
       }
-    } else None)
+    }
+  }
 
-    rdd.foreach {
-      case null =>
-      case r: RDD[_] =>
+  @tailrec
+  private def clearPartitions(rdds: Seq[RDD[_]]): Unit = {
+    val children = rdds.flatMap {
+      case null => Nil
+      case r =>
         // f.set(r, null)
-        UnsafeHolder.getUnsafe.putObject(r, rdd_partitions_, null)
-    }
-    if (children.isEmpty) {
-      return
+        UnsafeHolder.getUnsafe.putObject(r, CachedDataFrame.rdd_partitions_, null)
+        r.dependencies.map(_.rdd)
     }
-    clearPartitions(children)
+    if (children.nonEmpty) clearPartitions(children)
   }
 
   private def setPoolForExecution(): Unit = {
-    var pool = sparkSession.asInstanceOf[SnappySession].
-      sessionState.conf.activeSchedulerPool
-
+    var pool = snappySession.sessionState.conf.activeSchedulerPool
     // Check if it is pruned query, execute it automatically on the low latency pool
-    if (isLowLatencyQuery && shuffleDependencies.length == 0 && pool == "default") {
-      if (sparkSession.sparkContext.getAllPools.exists(_.name == "lowlatency")) {
-        pool = "lowlatency"
+    if (isLowLatencyQuery && shuffleDependencies.isEmpty && pool == "default") {
```

Comment:
It first reads the "activeSchedulerPool" so that will honor any explicit setting. Did I miss the question?

Reply:
isLowLatencyQuery call itself might return "true" or "false" depending on "cachedRDD.getNumPartitions".

So I thought this can change based on pruning and hence different pool can be used for different execution on the same CDF? 
Because if this won't change then perhaps 'lazy val' is fine instead of 'def' i.e. evaluating again on every execution.

Reply:
Right now won't change but can in future. The evaluation is very cheap so doesn't matter. Also if made lazy val then advantage can only be if it is copied over to duplicated CDF like done with other expensive fields. Doesn't look like worthwhile saving.

Reply:
That's fine here as cost is not high. Was thinking more in terms of whether it should be a constant or not?